### PR TITLE
Experiment with .preserve = FALSE as default

### PR DESCRIPTION
**Pull request by romainfrancois**

### CODE REVIEW

**Comment by romainfrancois** [comment]
closes #4002

File: R/compat-R.R [hunk 1]
```diff
@@ -0,0 +1,5 @@
+if(getRversion() < "3.2.0") {
+  lengths <- function(x, use.named = TRUE) {
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/compat-R.R]: Thanks

File: R/tbl-df.r [hunk 2]
```diff
@@ -42,8 +42,26 @@ arrange_.tbl_df <- function(.data, ..., .dots = list(), .by_group = FALSE) {
...
+  if (!.preserve && is_grouped_df(.data)) {
+    # we only need to do a group_by on the grouping data
+    # TODO: there's probably a way to make this more efficient internally
+    group_vars <- group_vars(out)
+    gd <- attr(out, "groups")
+    filtered <- group_data(group_by_at(
+      select_at(filter(gd, lengths(.rows) > 0, .preserve = FALSE), group_vars),
+      group_vars
+    ))
+    joined <- left_join(select(filtered, -last_col()), gd, by = group_vars)
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/tbl-df.r]: For example in that test case: 

```r
test_that("filter and slice keep zero length groups", {
  expect_equal(group_size(filter(df, f == 1)), c(2, 0, 0) )
  expect_equal(group_size(slice(df, 1)), c(1, 1, 0) )
})
```

The `left_join` is used to get the right `.rows` that are related to the data. Anyway, there's a `TODO` in the comment, will revise later. 

```
# TODO: there's probably a way to make this more efficient internally
```

File: R/tbl-df.r [hunk 3]
```diff
@@ -42,8 +42,26 @@ arrange_.tbl_df <- function(.data, ..., .dots = list(), .by_group = FALSE) {
   arrange_impl(.data, dots, environment())
 }
 
+filter_regroup <- function(.data, out, .preserve = FALSE) {
+  if (!.preserve && is_grouped_df(.data)) {
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/tbl-df.r]: Restructured the code slightly. Ideally `filter_regroup` would be done internally

File: R/tbl-df.r [hunk 4]
```diff
@@ -54,31 +72,29 @@ filter.tbl_df <- function(.data, ..., .preserve = TRUE) {
...
+  out <- slice_impl(.data, quo)
+  filter_regroup(.data, out, .preserve)
 }
 #' @export
-slice_.tbl_df <- function(.data, ..., .dots = list()) {
-  dots <- compat_lazy_dots(.dots, caller_env(), ..., .named = TRUE)
-  slice_impl(.data, dots[[1L]])
+slice_.tbl_df <- function(.data, ..., .preserve = FALSE, .dots = list()) {
+  dots <- compat_lazy_dots(.dots, caller_env(), ...)
+  slice(.data, !!!dots, .preserve = .preserve)
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/tbl-df.r]: it's just so that everything goes through the same code base

[the remark this answers is not in the document]
Comment (krlmlr) [reply R/tbl-df.r]: Maybe still revert.

[the remark this answers is not in the document]
Comment (krlmlr) [reply R/tbl-df.r]: We're already inconsistent in `slice_.data.frame()` vs. `slice_.tbl_df()`. Leaving it as is.

File: R/zzz.r [hunk 5]
```diff
@@ -6,6 +8,14 @@
   toset <- !(names(op.dplyr) %in% names(op))
   if (any(toset)) options(op.dplyr[toset])
 
+  if (getRversion() < "3.2.0") {
+    lengths <<- function(x, use.named = TRUE) {
```
Comment (krlmlr) [inline R/zzz.r]:
```suggestion
    lengths <<- function(x, use.names = TRUE) {
```

**Comment by krlmlr** [comment]
I had whipped up a slimmer R implementation:

```r
filter_regroup <- function(.data, out) {
  # we only need to do a group_by on the grouping data
  group_vars <- group_vars(out)

  # we keep only those rows in the grouping data that actually have rows
  # in the result
  gd <- filter(attr(out, "groups"), lengths(.rows) > 0L)

  # we keep those rows, the summarise() will add all missing rows
  # according to the new grouping
  empty_rows_value <- list(integer())

  new_gd <- ungroup(summarise(
    group_by_at(gd, group_vars),
    .rows = if (length(.rows)) .rows else empty_rows_value
  ))

  attr(out, "groups") <- new_gd
  out
}
```

Just in case you want to consider it. (Not sure what happens if there's a column called `.rows` in the data...)

File: R/zzz.r [hunk 6]
```diff
@@ -1,4 +1,6 @@
-lengths <- function(...) {}
+lengths <- function(x, use.names = TRUE) {
+  map_int(x, length)
```
Comment (krlmlr) [inline R/zzz.r]: `lengths()` is much better for R >= 3.2. Please see `R/lengths.R` and `R/zzz.R` in tibble.

Comment (romainfrancois) [reply R/zzz.r]: I don't find these files. Is this in a branch ?

Comment (krlmlr) [reply R/zzz.r]: Sorry, yes. tidyverse/tibble#551.

Comment (krlmlr) [reply R/zzz.r]: Scratch that. I'll find it.

Comment (romainfrancois) [reply R/zzz.r]: 😆 when you do that perhaps you can track other uses of `map_int + length`

Comment (krlmlr) [reply R/zzz.r]: It's in _pillar_. I can only look tomorrow.

**Pull request merged**